Undo/COUndoTrackStore.m

```diff
@@ -283,55 +402,180 @@ - (id) deserialize: (NSData *)data
 {
-	[_db executeUpdate: @"DELETE FROM commands WHERE uuid = ?", [aUUID dataValue]];
+	assert(dispatch_get_current_queue() != _queue);
+	
+	dispatch_sync(_queue, ^() {
+		[_db executeUpdate: @"DELETE FROM commands WHERE uuid = ?", [aUUID dataValue]];
+	});
 }
 
 - (BOOL) string: (NSString *)aString matchesGlobPattern: (NSString *)aPattern
 {
-	return [_db boolForQuery: @"SELECT 1 WHERE ? GLOB ?", aString, aPattern];
+	assert(dispatch_get_current_queue() != _queue);
+	__block BOOL result = NO;
+
+	dispatch_sync(_queue, ^() {
+		result = [_db boolForQuery: @"SELECT 1 WHERE ? GLOB ?", aString, aPattern];
+	});
+	
+	return result;
+}
+
+- (void)markCommandsAsDeletedForUUIDs: (NSArray *)UUIDs
+{
+	assert(dispatch_get_current_queue() != _queue);
+
+	// If there is a transaction underway, wait until it is finished
+	dispatch_semaphore_wait(_transactionLock, DISPATCH_TIME_FOREVER);
+	
+	__block NSArray *compactedTrackNames = nil;
+
+	// This can be run in background
+	dispatch_sync(_queue, ^() {
+		[_db beginTransaction];
+
+		for (ETUUID *UUID in UUIDs)
+		{
+			[_db executeUpdate: @"UPDATE commands SET deleted = 1 WHERE uuid = ?", [UUID dataValue]];
+		}
+
+		compactedTrackNames =
+			[_db arrayForQuery: @"SELECT DISTINCT trackname FROM commands WHERE deleted = 1 "];
+
+		[_db commit];
+	});
+
+	/* This must be run in the main thread:
+	   - no one must access _modifiedTrackStateForTrackName at the same time
+	   - we must post the commit notification immediately
+	   
+	   If we post it with some delay, someone could touch 
+	   _modifiedTrackStateForTrackName and overwrite COUndoTrackState.compacted. */
+	dispatch_sync_now(dispatch_get_main_queue(), ^() {
+		dispatch_sync(_queue, ^() {
+			[_db beginTransaction];
+
+			for (NSString *trackName in compactedTrackNames)
+			{
+				if (_modifiedTrackStateForTrackName[trackName] == nil)
+				{
+					_modifiedTrackStateForTrackName[trackName] = [self stateForTrackNameInCurrentQueue: trackName];
+				}
+				((COUndoTrackState *)_modifiedTrackStateForTrackName[trackName]).compacted = YES;
+			}
+
+			[_db commit];
+		});
+
+		[self postCommitNotifications];
+	});
+
+	dispatch_semaphore_signal(_transactionLock);
+}
+
+- (void)finalizeDeletions
+{
+	assert(dispatch_get_current_queue() != _queue);
+
+	// If there is a transaction underway, wait until it is finished
+	dispatch_semaphore_wait(_transactionLock, DISPATCH_TIME_FOREVER);
+	dispatch_sync(_queue, ^() {
+		[_db executeUpdate: @"DELETE FROM commands WHERE deleted = 1"];
+	});
+	dispatch_semaphore_signal(_transactionLock);
+}
+
+/**
+ * We run vacuum with dispatch_sync() in a queue, so we can sure no other
+ * threads access the database when we invoke -vacuum in a background thread.
+ */
+- (BOOL)vacuum
+{
+    assert(dispatch_get_current_queue() != _queue);
+	__block BOOL success = NO;
+
+	// If there is a transaction underway, wait until it is finished
+	dispatch_semaphore_wait(_transactionLock, DISPATCH_TIME_FOREVER);
```

Comment:
If we do dispatch_async() without a completion handler for -vacuum then we cannot know when the vacuum ends, right? I mean, we want to be told about it in order to remove the spinning indicator in the UI.

For -finalizeDeletions and -marksCommandsAsDeletedForUUIDs:, in a similar manner we have code that cannot be executed until these methods are done.

I'd be in favor of removing the semaphore too. A good option could be to solve the problem in the same way COSQLiteStore does it. By introducing a COUndoTrackStoreTransaction class hierarchy, we could eliminate -[COUndoTrackStore begin/endTransaction] and thereby the transaction lock too. This solution might seem a bit heavy, but it could result in a cleaner code in COUndoTrack too.